crates/paralegal-policy/src/context.rs

```diff
@@ -462,60 +470,65 @@ impl Context {
-                                ctrl_id: src.ctrl_id,
-                                typ: f.into(),
-                            })
-                        }
-                        nodes
-                    });
-
-                // Special case if sink is callarg, callsite is also an influencer
-                let arg_cs =
-                    if let NodeType::CallArgument(DataSink::Argument { function, .. }) = src.typ {
-                        Some(Node {
-                            ctrl_id: src.ctrl_id,
-                            typ: function.into(),
-                        })
-                    } else {
-                        None
-                    };
-                Box::new(influencees.chain(arg_cs.into_iter()))
-            };
+        let get_influencees = |influncee_sinks: Box<
+            dyn Iterator<Item = &'a CallSiteOrDataSink>,
+        >| {
+            let influencees = influncee_sinks.flat_map(move |cs_ds| {
+                // We definitely influence to the node itself.
+                let mut nodes = vec![Node {
+                    ctrl_id: src.ctrl_id,
+                    typ: cs_ds.into(),
+                }];
+                // If the node is a DataSink::Argument, we influence it's corresponding CallSite as well.
+                if let CallSiteOrDataSink::DataSink(DataSink::Argument { function: f, .. }) = cs_ds
+                {
+                    nodes.push(Node {
+                        ctrl_id: src.ctrl_id,
+                        typ: f.into(),
+                    })
+                }
+                nodes
+            });
+
+            // Special case if sink is callarg, callsite is also an influencer
+            let arg_cs =
+                if let NodeType::CallArgument(DataSink::Argument { function, .. }) = src.typ {
+                    Some(Node {
+                        ctrl_id: src.ctrl_id,
+                        typ: function.into(),
+                    })
+                } else {
+                    None
+                };
+            Box::new(influencees.chain(arg_cs.into_iter()))
+        };
 
         match edge_type {
-            EdgeType::Data => get_influencees(&self.flows_to[cf_id].data_flows_to),
-            EdgeType::DataAndControl => get_influencees(&self.flows_to[cf_id].flows_to),
+            EdgeType::Data => get_influencees(Box::new(
+                self.flows_to[&cf_id]
+                    .data_flows_to
+                    .row_set(&src_datasource.to_index(&self.flows_to[&cf_id].sources))
+                    .iter(),
+            )),
+            EdgeType::DataAndControl => get_influencees(Box::new(
+                self.flows_to[&cf_id]
+                    .sinks
+                    .as_vec()
+                    .iter()
+                    .filter(move |sink| {
+                        self.flows_to[&cf_id.clone()]
+                            .data_and_control_flows_to(src_datasource.clone(), (*sink).clone())
```

Comment:
This seems quite inefficient. The `data_and_control_flows_to` function executes a BFS over the graph and it will discover all influencees in one such pass but here we're calling that in a loop which has complexity `n*n` instead of `n`.

Reply:
One way to resolve this would be to instead create a method that returns an iterator of all data and control infuencees. We could use that iterator here and for `data_and_control_flow` we could call `any` on the iterator to test if the sink in question is contained.

And when I say create an iterator I mean a new struct that implements `Iterator` and where the `next` advances the BFS, yielding the nodes it discovers.

Reply:
done